### PR TITLE
Respond to all changes

### PR DESCRIPTION
**Pull request by lugray**
`changed_attributes` only responds to changes made by setter, missing those made by `write_attribute` directly, which includes changes to `updated_at` made by `touch`.  This adds and uses `all_changed_attributes` which corrects for this.

### CODE REVIEW

**Comment by lugray** [comment]
@dylanahsmith I've got tests passing.  In addition to targeting <=4.0 separately, I had to make a small change to your `test_transaction_changes_for_touch` to add
```
if ActiveRecord::VERSION::MAJOR < 4
      @user.save!
    end
```
after `@user.touch` because apparently that wasn't happening as part of `touch`.

File: ar_transaction_changes.gemspec [hunk 1]
```diff
@@ -2,6 +2,7 @@
 lib = File.expand_path('../lib', __FILE__)
 $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
 require 'ar_transaction_changes/version'
+require 'all_changed_attributes/version'
```
Comment (dylanahsmith) [inline ar_transaction_changes.gemspec]: This version isn't even used in this gemspec.  Why did you add it in the first place?

File: lib/all_changed_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+require 'all_changed_attributes/version'
+
+module AllChangedAttributes
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: All constants should be nested under the gem's top-level constant ArTransactionChanges to avoid polluting the global namespace.

File: lib/all_changed_attributes.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  def write_attribute(attr_name, value)
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: This method should stay private

File: lib/all_changed_attributes.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    store_all_changed_attributes(attr_name, value) if attributes[attr_name] != value
+    super
+  end
+
+  def all_changed_attributes(new = {})
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: I think the `new` argument would be confusing for a caller.  I think you should just change `all_changed_attributes(...)` calls into `all_changed_attributes.merge(...)` if they want to merge in attributes

File: lib/all_changed_attributes.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    new.merge(@all_changed_attributes ||= {})
+  end
+
+  def store_all_changed_attributes(attr_name, value)
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: The method name is `store_all_changed_attributes` but it just stores a single attribute.  This method also seems private

File: lib/all_changed_attributes.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    @all_changed_attributes = all_changed_attributes({attr_name.to_s => attributes[attr_name]})
+  end
+
+  if ActiveRecord::VERSION::MAJOR > 4 || (ActiveRecord::VERSION::MAJOR == 4 && ActiveRecord::VERSION::MINOR > 0)
+    def changes_applied
+      @all_changed_attributes = {}
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: This and the following method patches wouldn't me needed if we didn't expose `all_changed_attributes` as a separate feature, since the ArTransactionChanges module already clears `@transaction_changed_attributes` at the end of the transaction

File: lib/all_changed_attributes.rb (hunk 5 above)
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: The `value` parameter isn't used.  Instead you are using `attributes[attr_name]`

File: lib/all_changed_attributes.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    store_all_changed_attributes(attr_name, value) if attributes[attr_name] != value
+    super
+  end
+
+  def all_changed_attributes(new = {})
+    new.merge(@all_changed_attributes ||= {})
+  end
+
+  def store_all_changed_attributes(attr_name, value)
+    @all_changed_attributes = all_changed_attributes({attr_name.to_s => attributes[attr_name]})
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: If an attribute is changed multiple times in a transaction, then it looks like this won't store the original value, it will store the second last value.

File: test/transaction_changes_test.rb [hunk 8]
```diff
@@ -66,4 +66,27 @@ def test_transaction_changes_after_reload_unsaved
...
+    assert_equal [old_updated_at, @user.updated_at], @user.stored_transaction_changes["updated_at"]
+  end
+
+  def test_transaction_changes_for_touch
+    @user.update_attributes!(updated_at: Time.now - 1.second)
+    old_updated_at = @user.updated_at
+    @user.stored_transaction_changes = nil
+
+    @user.touch
+    if ActiveRecord::VERSION::MAJOR < 4
```
Comment (dylanahsmith) [inline test/transaction_changes_test.rb]: I think we can just remove support for rails 3.2 now.  I think this gem is mostly used for identity_cache and it requires `activerecord >= 4.0.4`

File: lib/all_changed_attributes.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  if ActiveRecord::VERSION::MAJOR > 4 || (ActiveRecord::VERSION::MAJOR == 4 && ActiveRecord::VERSION::MINOR > 0)
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: It looks like rails only supports 4.2.x and 5.0.x now (http://guides.rubyonrails.org/maintenance_policy.html) so maybe we can drop support for these earlier versions from ar_transaction_changes and identity_cache.

Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: I've dropped support for these old versions of rails from ar_transaction_changes and identity_cache

File: ar_transaction_changes.gemspec (hunk 1 above)
Comment (lugray) [inline ar_transaction_changes.gemspec]: Only because I was following your example: this is my first week at Shopify outside of onboarding, and I have no previous ruby experience.  I'll remove it.

File: lib/all_changed_attributes.rb (hunk 6 above)
Comment (lugray) [inline lib/all_changed_attributes.rb]: Without this the double save test fails.

**Comment by lugray** [comment]
@maartenvg Could you take a look at this too and see what you think?

**Comment by lugray** [comment]
Could I get some 👀 on this @maartenvg and @dylanahsmith?

File: lib/all_changed_attributes.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
+module ArTransactionChanges
+  module AllChangedAttributes
+    def all_changed_attributes
+      @all_changed_attributes ||= ActiveSupport::HashWithIndifferentAccess.new
+    end
+
+    private
+
+    def store_to_all_changed_attributes(attr_name)
+      @all_changed_attributes = {attr_name.to_s => attributes[attr_name]}.merge(all_changed_attributes)
```
Comment (dylanahsmith) [inline lib/all_changed_attributes.rb]: This will create two ruby hash objects for each write.  In this case you just can modify the existing hash if it doesn't already have the key.

File: lib/attributes_changed_by_write.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
+module ArTransactionChanges
+  module AttributeChangedByWrite
+    def attributes_changed_by_write
+      @attributes_changed_by_write ||= ActiveSupport::HashWithIndifferentAccess.new
```
Comment (dylanahsmith) [inline lib/attributes_changed_by_write.rb]: Continuing the discussion from https://github.com/dylanahsmith/ar_transaction_changes/pull/10#discussion_r72691942 we should just use `@transaction_changed_attributes` and just change the way it is being set to capture all attributes.

You said doing that causes the double save test to fail, but I tried doing that and the tests pass.  See https://github.com/dylanahsmith/ar_transaction_changes/commit/4888ac82dcce02b9428aa66134ca1dd1feab982a for reference
